perun-registrar-lib/src/main/java/cz/metacentrum/perun/registrar/impl/MailManagerImpl.java

```diff
@@ -1091,69 +1086,61 @@ private void setFromAndReplyTo(MimeMessage message, String fromName, String from
-	 * @return list of mail addresses to send mail to
+	 * @return Set of mail addresses to send mail to
 	 */
-	private List<String> getToMailAddresses(Application app) {
-		List<String> result = new ArrayList<>();
-
-		// get proper value from attribute
+	private Set<String> getToMailAddresses(Application app) {
+		Set<String> result = new HashSet<>();
 		try {
-			Attribute attrSenderEmail;
-
-			if (app.getGroup() == null) {
-				// it is a VO application
-				return getMailsFromVo(app, URN_VO_TO_EMAIL);
-			}
-
-			attrSenderEmail = attrManager.getAttribute(registrarSession, app.getGroup(), URN_GROUP_TO_EMAIL);
-
-			if (attrSenderEmail == null || attrSenderEmail.getValue() == null) {
-				// not specified toEmail group attribute
-
-				// try to use all group admins with specified preferred emails
-				List<String> rolesToNotify = getNotificationReceiverRoles(GROUP);
-				List<RichUser> admins = new ArrayList<>();
-				for (String role : rolesToNotify) {
-					admins.addAll(AuthzResolverBlImpl.getRichAdmins(registrarSession, app.getGroup(), role));
-				}
-				result.addAll(getUserPreferredMails(registrarSession, admins));
-				result = result.stream().distinct().toList();
-
-				if (!result.isEmpty()) {
-					return result;
-				}
-				return getMailsFromVo(app, URN_VO_TO_EMAIL);
+			// Check group first
+			if (app.getGroup() != null) {
+				// we have group application
+				Group currentGroup = app.getGroup();
+				do {
+					result = getToEmailsGroupLevel(currentGroup);
+					if (!result.isEmpty()) {
+						break;
+					} else {
+						currentGroup = getParentGroupForMailAddresses(currentGroup);
+					}
+				} while (currentGroup != null);
 			}
-
-			// specified toEmail group attribute, use valid emails
-			ArrayList<String> value = attrSenderEmail.valueAsList();
-			for (String adr : value) {
-				if (adr != null && !adr.isEmpty()) {
-					result.add(adr);
-				}
+			// VO application or fallback to VO level
+			if (app.getGroup() == null || result.isEmpty()) {
+				result = getToEmailsVoLevel(app.getVo());
 			}
-
 		} catch (Exception ex) {
 			// we don't care about exceptions here - we have backup TO/FROM address
-			if (app.getGroup() == null) {
-				log.error("[MAIL MANAGER] Exception thrown when getting TO email from an attribute {}. Ex: {}", URN_VO_TO_EMAIL, ex);
-			} else {
-				log.error("[MAIL MANAGER] Exception thrown when getting TO email from an attribute {}. Ex: {}", URN_GROUP_TO_EMAIL, ex);
-			}
+			log.error("[MAIL MANAGER] Exception thrown when getting TO email from an attribute {}. Ex: {}",
+					app.getGroup() == null ? URN_VO_TO_EMAIL : URN_GROUP_TO_EMAIL, ex);
 			// set backup
 			result.clear();
+		}
+		// no recipients found for Group and Vo, use backup
+		if (result.isEmpty()) {
 			result.add(getPropertyFromConfiguration("backupTo"));
 		}
-
 		return result;
 	}
 
+	private Group getParentGroupForMailAddresses(Group group) {
+		if (group != null && group.getParentGroupId() != null) {
```

Comment:
I don't think members should be checked specifically. IMHO:
1) members group cannot have subgroups
2) members group is just a representation of VO, so technically we are interested in VO settings instead. 

But if you wish, I can modify it.

Reply:
What I meant is that `getParentGroup()` returns the `members` group when the passed group does not have a parent.
It would be more confusing probably, leave it like this.

Reply:
Ah, got it!